Content.Server/VendingMachines/VendingMachineSystem.cs

```diff
@@ -0,0 +1,243 @@
+using Robust.Shared.Audio;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Player;
+using System;
+using System.Linq;
+using Content.Server.Popups;
+using Content.Server.Power.Components;
+using Content.Server.WireHacking;
+using Content.Shared.Access.Components;
+using Content.Shared.Access.Systems;
+using Content.Shared.VendingMachines;
+using Robust.Server.GameObjects;
+using Robust.Shared.Localization;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+using Content.Server.Advertise;
+using Content.Server.Throwing;
+using Robust.Shared.Maths;
+using static Content.Shared.VendingMachines.SharedVendingMachineComponent;
+
+namespace Content.Server.VendingMachines.systems
+{
+    public sealed class VendingMachineSystem : EntitySystem
+    {
+        [Dependency] private readonly AdvertiseSystem _advertisementSystem = default!;
+        [Dependency] private readonly IRobustRandom _random = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<VendingMachineComponent, ComponentInit>(OnComponentInit);
+            SubscribeLocalEvent<VendingMachineComponent, ActivateInWorldEvent>(HandleActivate);
+            SubscribeLocalEvent<VendingMachineComponent, PowerChangedEvent>(OnPowerChanged);
+        }
+
+        private void OnComponentInit(EntityUid uid, VendingMachineComponent component, ComponentInit args)
+        {
+            base.Initialize();
+
+            if (component.UserInterface != null)
+            {
+                component.UserInterface.OnReceiveMessage += component.OnUiReceiveMessage;
+            }
+            if (EntityManager.TryGetComponent(component.Owner, out ApcPowerReceiverComponent? receiver))
+            {
+                TryUpdateVisualState(component, receiver.Powered ? VendingMachineVisualState.Normal : VendingMachineVisualState.Off);
+            }
+
+            InitializeFromPrototype(component);
+        }
+
+        private void HandleActivate(EntityUid uid, VendingMachineComponent component, ActivateInWorldEvent args)
+        {
+            if (!component.Powered ||
+                !EntityManager.TryGetComponent(args.User, out ActorComponent? actor))
+            {
+                return;
+            }
+            var wires = EntityManager.GetComponent<WiresComponent>(component.Owner);
+            if (wires.IsPanelOpen)
+            {
+                wires.OpenInterface(actor.PlayerSession);
+            } else
+            {
+                component.UserInterface?.Toggle(actor.PlayerSession);
+            }
+        }
+
+        private void OnPowerChanged(EntityUid uid, VendingMachineComponent component, PowerChangedEvent args)
+        {
+            TryUpdateVisualState(component);
+        }
+
+        public void InitializeFromPrototype(VendingMachineComponent component)
```

Comment:
Here and bellow, use standard systems public functions convention `InitializeFromPrototype(uid, component?, ...)`. Read more about [here ](https://docs.spacestation14.io/en/getting-started/conventions) in "Public API Method Signature"